sound/soc/sof/topology.c

```diff
@@ -1364,6 +1367,42 @@ static int sof_widget_load_dai(struct snd_soc_component *scomp, int index,
 		memcpy(&dai->comp_dai, &comp_dai, sizeof(comp_dai));
 	}
 
+	if (comp_dai.type != SOF_DAI_INTEL_ALH)
+		goto OUT;
+
+	/*
+	 * Some ALH DAIs are not configured by sof_link_alh_load(), so
+	 * configure them here
+	 */
```

Comment:
?? I really wasn't able to understand this and why we need to have ALH code in here. That looks rather bad to be honest.

Reply:
@plbossart I must admit that is a little hacky. The issue is that ``sof_link_load() -> sof_link_alh_load()`` will only be called when a FE link is created. But in our multi cpu dai link case , there are two FE DAIs in a link. As a result the second DAI will not be configured. Maybe a better solution is to move all DAI config from sof_link_load() to sof_widget_load_dai() since it is "DAI" config not "link" config.

Reply:
But all the DAI_CONFIG we have in topology is actually dai-link config...
I guess I need a lot more coffee today, or call the cavalry. @ranj063 what do you think?

Reply:
@bardliao @plbossart sof_link_load() is called every FE and BE DAI links. I think whats missing is something like what we do in sof_link_hda_process() for all DAIs with the same name as the link name. We should not be doing this here.

Reply:
@ranj063 but this case is different from HDaudio, we have multiple cpu- and codec-dais per link, so a name-based solution cannot possibly work, can it?

Reply:
@plbossart if thats the case then why do we call sof_set_dai_config() in sof_link_alh_load()? 
As far as I can tell, it uses the name based matching too.

Reply:
> But all the DAI_CONFIG we have in topology is actually dai-link config...

@plbossart Yes, and that't the problem. Why should we set dai config data in dai-link config?
It works with an assumption that there is only one dai connected to a dai link.  My point is that the dai config data such as SSP_CONFIG_DATA or ALH_CONFIG_DATA in our tplg files is actually for DAI not for link. See also https://github.com/thesofproject/sof/pull/2692


> @bardliao @plbossart sof_link_load() is called every FE and BE DAI links. I think whats missing is something like what we do in sof_link_hda_process() for all DAIs with the same name as the link name. We should not be doing this here.

@ranj063 @plbossart The question is where should we get the DAI config or are we allowed to have different dai configs to different dais in the same link? 

> @plbossart if thats the case then why do we call sof_set_dai_config() in sof_link_alh_load()?

That may be a mistake. Currently, sof_set_dai_config() will set the dai config to all the dais which match to the link name, but only one of their dai config has been sent to FW via ipc.

Reply:
@plbossart @ranj063  https://github.com/thesofproject/linux/pull/1982 may be a solution If we can assume that all DAIs in the same link have the same config.

Reply:
> @ranj063 @plbossart The question is where should we get the DAI config or are we allowed to have different dai configs to different dais in the same link?
> 
> >

@bardliao Im not sure how you could have different config for DAIs on the same DAI link.

Reply:
the slot information could be different.

Reply:
@plbossart but then the topology parsing logic today is flawed isnt it? We assume all DAIs with the same name as the DAI link have the same config. There's no way to differentiate them today

Reply:
My remark was a bit hypothetical. But it's possible that in the future we have a cpu-dai configuration to use Lane0 on one segment and Lane1 on the other. Of if we have a dailink with two synchronized SSPs (possible with TGL+), that we use different slots for different cpu-dais, specifically on capture.
I think everything from sample rate to number of channels will typically be the same, it's the link configurations that could vary.
I am a bit of the fence here on whether to deal with this now. maybe we just make this assumption for now, document it explicitly and think later about how we might want to sort this out. Unless you have a solution in the next day or so it may be the right path forward.

Reply:
@plbossart is it only the SSPs that would have the need for different slot info? Or do we envision other types of DAIs to also have different config for the different DAIs

Reply:
If it only the SSPs, then can we possibly think of specifying the different hw_config for different DAIs and use the ID info in the hw_config to match which DAI we want the hw_config to be applicable to instead of the default_hw_config. Just thinking out loud.

Reply:
> @bardliao Im not sure how you could have different config for DAIs on the same DAI link.

What about a link with subwoofer amp?  Something like 3 dais, two of them are 2 channels (normal amp) and one of them is 1 channel (subwoofer amp). We don't have this feature for now, but maybe our DSP can do it one day. 

> @plbossart but then the topology parsing logic today is flawed isnt it?

I have to say yes, unfortunately. That's why I was thinking if we should move xxx_CONFIG_DATA from DAI_CONFIG to DAI_ADD in the topology file and do the corresponding change on kernel side. But it will be a big change. So I prefer to make an assumption that all dais in the same link have the same config now and do the change when we really need it.

Reply:
@bardliao OK lets file an enhancement issue to track this as well